libs/small-steps/src/Data/Sharing.hs

```diff
@@ -0,0 +1,184 @@
+{-# LANGUAGE BangPatterns #-}
+{-# LANGUAGE DataKinds #-}
+{-# LANGUAGE FlexibleInstances #-}
+{-# LANGUAGE GADTs #-}
+{-# LANGUAGE KindSignatures #-}
+{-# LANGUAGE LambdaCase #-}
+{-# LANGUAGE MultiParamTypeClasses #-}
+{-# LANGUAGE ScopedTypeVariables #-}
+{-# LANGUAGE TypeApplications #-}
+
+module Data.Sharing
+  ( emptyA,
+    include,
+    internI,
+    internUsingMap,
+    FromCBORShare (..),
+    Intern (..),
+    fromCBOR',
+    -- Extra for experts
+    Arity (..),
+    Nat (..),
+    IList (..),
+  )
+where
+
+import Cardano.Binary (Decoder, FromCBOR (..), decodeListLen, decodeMapSkel, dropMap)
+import Control.Iterate.SetAlgebra (BiMap, biMapFromAscDistinctList)
+import Control.Monad (void)
+import Control.SetAlgebra (backwards)
+import Data.Coders (decodeMap, decodeVMap, invalidKey)
+import Data.Compact.VMap (VB, VMap)
+import qualified Data.Compact.VMap as VMap
+import qualified Data.List as List
+import qualified Data.Map.Strict as Map
+import Data.Map.Strict.Internal
+
+-- =======================================
+
+class Intern map where
+  findKey :: Ord k => k -> map k v -> Maybe k
+  internSize :: map k v -> Int
+
+-- | An Intern list
+data IList k where
+  INil :: IList k
+  ICons :: Intern map => map k a -> IList k -> IList k
+
+instance Show k => Show (IList k) where
+  show INil = "INil"
+  show (ICons x xs) = show (internSize x) ++ " " ++ show xs
+
+-- ==================================
+
+internUsingMap :: Ord k => k -> Map k a -> k
+internUsingMap = go
+  where
+    go !k Tip = k
+    go !k (Bin _ kx _ l r) = case compare k kx of
+      LT -> go k l
+      GT -> go k r
+      EQ -> kx
+
+instance Intern Map where
+  findKey = go
+    where
+      go _ Tip = Nothing
+      go !k (Bin _ kx _ l r) = case compare k kx of
+        LT -> go k l
+        GT -> go k r
+        EQ -> Just kx
+  internSize m = Map.size m
+
+-- | Intern a key, return an Eq key, which is the same Heap object if it occurs in the IList
+--   Sharing aside, (internI ilist) is the identity function.
+internI :: Ord k => IList k -> k -> k
+internI INil k = k
+internI (ICons m ms) k =
+  case findKey k m of
+    Just k2 -> k2
+    Nothing -> internI ms k
+
+-- | include a new map in the IList, the idea is to keep the IList sorted by map size, so
+--   the largest maps come first. This should decrease the number of maps searched when
+--   looking to intern a 'k' object.
+include :: Intern map => map k a -> IList k -> IList k
+include m INil = ICons m INil
+include m (x@(ICons n ns)) =
+  case compare (internSize m) (internSize n) of
+    EQ -> ICons n (include m ns) -- When the size is the same, keep the older map at the front
+    GT -> ICons m x
+    LT -> ICons n (include m ns)
```

Comment:
A bit of simplification
```suggestion
include m x@(ICons n ns)
  | internSize m > internSize n = ICons m x
  | otherwise = ICons n (include m ns)
```